arangod/StorageEngine/TransactionState.cpp

```diff
@@ -82,9 +84,7 @@ TransactionCollection* TransactionState::collection(
   TRI_ASSERT(_status == transaction::Status::CREATED ||
              _status == transaction::Status::RUNNING);
 
-  std::lock_guard lock(_collectionsLock);
   auto collectionOrPos = findCollectionOrPos(cid);
```

Comment:
We should rather acquire the read lock here, instead of acquring it inside `findCollectionOrPos()`. `findCollectionOrPos(...)` is protected (but could even be made private).
That probably then only leaves the ReplicatedRocksDBTranactionState::doCommit()` method requiring the lock recursively, which could also easily be fixed. And then we wouldn't need a recursive lock anymore.

Reply:
Unfortunately, we need to use a recursive lock in `TransactionState::addCollectionInternal` regardless of the `findCollectionOrPos` implementation, because somewhere down the line it gets into `RocksDBTransactionState::isOnlyExclusiveTransaction`, which again has to iterate over the list of collections. So there would be at least one other method which needs the recursive lock.
I have tried a minimalistic approach in https://github.com/arangodb/arangodb/pull/19670/commits/b8f3b9fd638cd22d6db099b5dc68cf99dd0e70d2, but then I followed with a "better safe than sorry" approach.